### DISCUSSION TITLE
NEXTAUTH_URL is redirecting to a different uri than my github app callback URL

### DISCUSSION THREAD

**Discussion by nicolasguasca1** [vercel/community#697] [Help] [answered]
Hello guys, Thank you for opening this space for discussions. I have the following situation. Steps to reproduce:
1. Set a .env.local file with the following value: NEXTAUTH_URL="http://subdomain.localhost:3000".
2. Set a callback url on a Github app with this value: "http://subdomain.localhost:3000".
3. Create a middleware function to deal with the rewriting of the subdomain like this:

```import { NextResponse } from "next/server";

import type { NextRequest } from "next/server";
import { request } from "http";

export default function middleware(req: NextRequest) {
  const newReq = req.nextUrl.clone();
  const regex = new RegExp(
    `*********\.vercel\.app`
  );

  const pathname = req.nextUrl.pathname.toString();

  const hostname = req.headers.get("host");

  if (!hostname)
    return new Response(null, {
      status: 400,
      statusText: "No hostname found in request headers"
    });

  const currentHost =
    process.env.VERCEL_ENV === `production` ||
    process.env.VERCEL_ENV === `preview`
      ? 
        hostname
          .replace(`${process.env.VERCEL_URL}`, "")
          .replace(`${process.env.NEXT_PUBLIC_VERCEL_URL}`, "")
      : hostname.replace(`.localhost:3000`, "");

  if (pathname.startsWith(`/_sites`))
    return new Response(null, {
      status: 404
    });

  if (
    !pathname.includes(".") &&
    !pathname.startsWith("/api")
  ) {
    if (currentHost === "subdomain") {
      if (
        pathname === "/login" &&
        (req.cookies["next-auth.session-token"] ||
          req.cookies["__Secure-next-auth.session-token"])
      ) {
        newReq.pathname = "/";
        return NextResponse.redirect(newReq.toString());
      }

      newReq.pathname = `/subdomain${pathname}`;
      console.log(newReq.toString()); // ==>>>>>This logs: http://localhost:3000/subdomain/login
      return NextResponse.rewrite(newReq.toString());
    }

    newReq.pathname = `${pathname}`;
    return NextResponse.rewrite(newReq.toString());
  }
}
```

Expected behaviour:

- The next app uses the session token to redirect back to `/` or ask for the credentials and then go to `/`.
- Actual behaviour: nextauth warns that the _redirect uri_ doesn't match the one specified in the _callback url_ of the github app (which are actually the same). 

Any help would be realy appreciated folks!

**Comment by amyegan** [comment, marked as answer]
_Solution shared in reply https://github.com/vercel/community/discussions/697#discussioncomment-3201871_

Hi @nicolasguasca1. I'm sorry for the late response. Are you still facing this issue, or have you found a solution already?

**Reply by nicolasguasca1** [reply]
No worries @amyegan ! I got it to work by setting the Github app callback value as `http://subdomain.localhost:3000/api/auth/callback/github` but what I am facing now is that based on the way NEXTAUTH is configured, the `providers` array has set the callback url for the github provider as `http://subdomain.localhost:3000/api/auth/callback/github` but I need to use SSL on my local because my intend is to integrate Facebook Provider. I am now unable to change the `callbackUrl` set by NEXTAUTH for the providers. I tryied several times on the callbacks config on my [...NEXTAUTH].ts file without sucess.

**Reply by amyegan** [reply]
I don't want you to feel like I'm giving you the runaround, but I think [NextAuth.js Discussions](https://github.com/nextauthjs/next-auth/discussions) is the right place to ask about this problem. They know a lot more than I do about how NextAuth.js works.

The only other tip I can think of right now, specific to Vercel, is to try [using your Vercel Project to manage environment variables](https://vercel.com/docs/concepts/projects/environment-variables) instead of a `.env` file. Then run the project with [`vercel dev`](https://vercel.com/docs/cli#commands/dev) in your local environment to pull in the environment variables for your Project. But I'm not sure if the environment variables are the real problem for you.
